src/Share/Survey/SurveyHelper.cs

```diff
@@ -0,0 +1,360 @@
+
+        private void MergeScheduleInfo(ScheduleInfo externalScheduleInfo)
+        {
+            DateTime externalLastPromptDate = Convert.ToDateTime(externalScheduleInfo?.LastPromptDate);
+            IDictionary<string, ModuleInfo> externalModules = new Dictionary<string, ModuleInfo>();
+            foreach(ModuleInfo info in externalScheduleInfo?.Modules)
+            {
+                externalModules[info.Name] = info;
+            }
+
+            HashSet<string> moduleNames = new HashSet<string>(Modules.Keys);
+            moduleNames.UnionWith(new HashSet<string>(externalModules.Keys));
+
+            foreach (string name in moduleNames)
+            {
+                if (externalModules.ContainsKey(name) && (!Modules.ContainsKey(name) || Convert.ToDateTime(Modules[name].LastActiveDate) < Convert.ToDateTime(externalModules[name].LastActiveDate)))
+                {
+                    if (externalLastPromptDate != DateTime.MinValue && Convert.ToDateTime(externalModules[name].LastActiveDate) == externalLastPromptDate)
+                    {
+                        LastPromptDate = externalLastPromptDate;
+                    }
+                    Modules[name] = new ModuleInfo(externalModules[name]);
+                }
+            }
+        }
+
+        private ScheduleInfo GetScheduleInfo()
+        { 
+            return new ScheduleInfo() { LastPromptDate = LastPromptDate.ToString("yyyy-MM-dd"), Modules = Modules.Values.ToList() };
+        }
+
+        private bool ReadFromStream()
+        {
+            StreamReader sr = null;
+            try
+            {
+                if (File.Exists(SurveyScheduleInfoFile))
+                {
+                    sr = new StreamReader(new FileStream(SurveyScheduleInfoFile, FileMode.Open, FileAccess.Read, FileShare.None));                    
+                    MergeScheduleInfo(JsonConvert.DeserializeObject<ScheduleInfo>(sr.ReadToEnd()));
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is UnauthorizedAccessException)
+                {
+                    this.IgnoreSchedule = true;
+                }
+                //deserialize failed, means content of file is incorrect, make file empty
+                if (e is JsonException)
+                {
+                    if (sr != null)
+                    {
+                        sr.Dispose();
+                    }
+                    EmptyFileAsync();
+                }
+                return false;
+            }
+            finally
+            {
+                if (sr != null)
+                {
+                    sr.Dispose();
+                }
+            }
+            return true;
+        }
+
+        private bool WriteToStream(string info)
+        {
+            StreamWriter sw = null;
+            try
+            {
+                sw = new StreamWriter(new FileStream(SurveyScheduleInfoFile, FileMode.Create, FileAccess.Write, FileShare.None));
```

Comment:
The read and write can be on different threads. With this, the following write will fail once a thread is reading/writing. This may not be a big deal. Just want to bring it to your attentions.

Reply:
Our strategy is to fail other read/write while one thread is writing.
Failures of read/write were handled and acceptable.